hare: enable cross-compilation

onemoresuza opened the pull request:
## Description of changes

- **qbe: enable cross-compilation**
- **harec: enable cross-compilation**
- **hare: enable cross-compilation**
- **haredoc: init at 0.24.0**
- Add mention to 24.05 release notes about the split of `hare` and `haredoc`.


<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

## Note for reviewers

To test this PR:

```bash
git clone --filter=blob:none --depth=1 -b hare-crosscomp https://github.com/onemoresuza/nixpkgs.git hare-crosscomp
cd hare-crosscomp
nix-build -E '
    let
        pkgs = import ./. {};
        x86_64 = builtins.attrValues {
            inherit (pkgs.pkgsCross.gnu64) qbe harec hare haredoc;
        };
        aarch64 = builtins.attrValues {
            inherit (pkgs.pkgsCross.aarch64-multiplatform) qbe harec hare haredoc;
        };
    in
        aarch64 ++ x86_64
'
```

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

onemoresuza committed:
qbe: enable cross-compilation

onemoresuza committed:
harec: enable cross-compilation

onemoresuza committed:
hare: enable cross-compilation

onemoresuza commented:
Result of `nixpkgs-review pr 292608` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>himitsu-firefox</li>
  </ul>
</details>
<details>
  <summary>16 packages built:</summary>
  <ul>
    <li>bonsai</li>
    <li>hare</li>
    <li>hare.man</li>
    <li>hareThirdParty.hare-compress</li>
    <li>hareThirdParty.hare-ev</li>
    <li>hareThirdParty.hare-json</li>
    <li>hareThirdParty.hare-png</li>
    <li>hareThirdParty.hare-ssh</li>
    <li>hareThirdParty.hare-toml</li>
    <li>harec</li>
    <li>haredo</li>
    <li>haredo.man</li>
    <li>himitsu</li>
    <li>qbe</li>
    <li>treecat</li>
    <li>treecat.man</li>
  </ul>
</details>

onemoresuza commented:
Result of `nixpkgs-review pr 292608` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>himitsu-firefox</li>
  </ul>
</details>
<details>
  <summary>16 packages built:</summary>
  <ul>
    <li>bonsai</li>
    <li>hare</li>
    <li>hare.man</li>
    <li>hareThirdParty.hare-compress</li>
    <li>hareThirdParty.hare-ev</li>
    <li>hareThirdParty.hare-json</li>
    <li>hareThirdParty.hare-png</li>
    <li>hareThirdParty.hare-ssh</li>
    <li>hareThirdParty.hare-toml</li>
    <li>harec</li>
    <li>haredo</li>
    <li>haredo.man</li>
    <li>himitsu</li>
    <li>qbe</li>
    <li>treecat</li>
    <li>treecat.man</li>
  </ul>
</details>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3553

pbsds reviewed (commented):
```
error: builder for '/nix/store/ws7l12hv17i4197kh5wiq60v24bcz04z-hare-aarch64-unknown-linux-gnu-0.24.0.drv' failed with exit code 1;
       last 10 log lines:
       > LD      .bin/hare
       > HARE    .bin/haredoc
       > .bin/hare: line 1: can't open �!��!�!@��Ң�B�A���: no such file
       > .bin/hare: line 2: can't open ���G: no such file
       > .bin/hare: line 3: @���Ң��B�A��ҡ��!�: not found
       > .bin/hare: line 3: can't open ��5�y��*q᧟�!S?q�T: no such file
       > : no such filee 4: can't open ���
       > .bin/hare: line 4: syntax error: unterminated quoted string
       > make: *** [Makefile:58: .bin/haredoc] Error 2
       > rm .harecache/dirs.s .harecache/hare_unparse.s .harecache/hare_lex.s .harecache/hash.s .harecache/strconv.s .harecache/path.s .harecache/unix.s .harecache/crypto_math.s .harecache/encoding_hex.s .harecache/getopt.s .harecache/fmt.s .harecache/strings.s .harecache/time_chrono.s .harecache/os_exec.s .harecache/sort_cmp.s .harecache/hare_parse.s .harecache/linux_vdso.s .harecache/time_date.s .harecache/errors.s .harecache/io.s .harecache/fs.s .harecache/cmd_hare_build.s .harecache/unix_signal.s .harecache/bufio.s .harecache/os.s .harecache/bytes.s .harecache/types_c.s .harecache/hare_ast.s .harecache/shlex.s .harecache/endian.s .harecache/crypto_sha256.s .harecache/memio.s .harecache/encoding_utf8.s .harecache/hare_module.s .harecache/time.s .harecache/math.s .harecache/cmd_hare.s .harecache/types.s .harecache/unix_tty.s .harecache/format_elf.s .harecache/ascii.s .harecache/linux.s .harecache/sort.s
```

onemoresuza commented on pkgs/by-name/ha/hare/package.nix:
```diff
@@ -32,6 +33,10 @@ in
+  qbePlatform =
+    if arch == "x86_64" then "amd64_sysv"
+    else if arch == "aarch64" then "arm64"
+    else "rv64";
```
Applied this patch.

onemoresuza commented:
> ```
> error: builder for '/nix/store/ws7l12hv17i4197kh5wiq60v24bcz04z-hare-aarch64-unknown-linux-gnu-0.24.0.drv' failed with exit code 1;
>        last 10 log lines:
>        > LD      .bin/hare
>        > HARE    .bin/haredoc
>        > .bin/hare: line 1: can't open �!��!�!@��Ң�B�A���: no such file
>        > .bin/hare: line 2: can't open ���G: no such file
>        > .bin/hare: line 3: @���Ң��B�A��ҡ��!�: not found
>        > .bin/hare: line 3: can't open ��5�y��*q᧟�!S?q�T: no such file
>        > : no such filee 4: can't open ���
>        > .bin/hare: line 4: syntax error: unterminated quoted string
>        > make: *** [Makefile:58: .bin/haredoc] Error 2
>        > rm .harecache/dirs.s .harecache/hare_unparse.s .harecache/hare_lex.s .harecache/hash.s .harecache/strconv.s .harecache/path.s .harecache/unix.s .harecache/crypto_math.s .harecache/encoding_hex.s .harecache/getopt.s .harecache/fmt.s .harecache/strings.s .harecache/time_chrono.s .harecache/os_exec.s .harecache/sort_cmp.s .harecache/hare_parse.s .harecache/linux_vdso.s .harecache/time_date.s .harecache/errors.s .harecache/io.s .harecache/fs.s .harecache/cmd_hare_build.s .harecache/unix_signal.s .harecache/bufio.s .harecache/os.s .harecache/bytes.s .harecache/types_c.s .harecache/hare_ast.s .harecache/shlex.s .harecache/endian.s .harecache/crypto_sha256.s .harecache/memio.s .harecache/encoding_utf8.s .harecache/hare_module.s .harecache/time.s .harecache/math.s .harecache/cmd_hare.s .harecache/types.s .harecache/unix_tty.s .harecache/format_elf.s .harecache/ascii.s .harecache/linux.s .harecache/sort.s
> ```

Was this error log produced by the `nix-build -A pkgsCross.aarch64-multiplatform.hare`?

pbsds commented:
yes

SuperSandro2000 commented:
@ofborg build pkgsCross.aarch64-multiplatform.hare

onemoresuza commented:
@pbsds, I am unable to reproduce the error encountered by both you and ofborg: Running `nix-build -A pkgsCross.aarch64-multiplatform.hare` on my `x86_64` machine does build the derivation with an `aarch64` executable as it's shown by running the `file` command on it:

```console
[user@hostname: ~/nixpkgs] $ file ./result/bin/.hare-wrapped
./result/bin/.hare-wrapped: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), statically linked, not stripped
```

pbsds commented:
Do you have `boot.binfmt.emulatedSystems` configured?

onemoresuza commented:
> Do you have `boot.binfmt.emulatedSystems` configured? That can affect the build and is regarded as impure/unreproducible

That was it! 

I was guessing it was something related to it since both you and ofborg only failed on `hare`, which means that the cross compilation of `qbe` and `harec` had succeeded --- I've also tested building them individually after disabling `binfmt`.

Regarding the error itself, it occurs because `haredoc`, Hare's documentation tool, is being built with the cross-compiled `hare` binary instead of the one from the buildPlatform. Going to take a look at [Void Linux's package](https://github.com/void-linux/void-packages/pull/48784/files), since [they mentioned](https://lists.sr.ht/~sircmpwn/hare-users/%3Cd62782ec-021e-4ee6-849b-ca11848f3a72%40fentker.eu%3E#%3Cd62782ec-021e-4ee6-849b-ca11848f3a72@fentker.eu%3E) that they had to patch the Makefile in regards to `haredoc` to make cross-compilation work.

onemoresuza committed:
haredoc: init at 0.24.0

Also add a mention to the release notes of 24.05 about `hare` and
`haredoc` being split into different packages.

onemoresuza commented:
I've patched `hare` so that it does not build `haredoc`, since the latter relies on the freshly build `$(BINOUT)/hare`, which makes cross-compilation not possible.

Now, `haredoc` is has its own package.

onemoresuza commented:
@ofborg build pkgsCross.aarch64-multiplatform.hare

pbsds commented:
Great work!

onemoresuza commented:
Thanks!
